internal/app/machined/pkg/system/services/registry/registry_test.go

```diff
@@ -145,3 +150,151 @@ func readAll(t *testing.T, resp *http.Response) string {
 
 	return builder.String()
 }
+
+//nolint:gocyclo
+func TestRegistryWithFakeFS(t *testing.T) {
+	cacheDir := t.TempDir()
+
+	logger := zaptest.NewLogger(t)
+
+	it := func(yield func(string) bool) {
+		if !yield(cacheDir) {
+			return
+		}
+	}
+
+	svc := registry.NewService(registry.NewMultiPathFS(it), logger)
+
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+
+	ctx, cancel := context.WithCancelCause(context.Background())
+	defer cancel(nil)
+
+	go func() {
+		defer wg.Done()
+
+		cancel(cmp.Or(svc.Run(ctx), errors.New("service exited")))
+	}()
+
+	defer wg.Wait()
+
+	type fakeRegistryData struct {
+		testName     string
+		registryName string
+		imageName    string
+		blobsData    map[string]string
+		tag          string
+	}
+
+	blobDir := filepath.Join(cacheDir, "blob")
+	manifestsDir := filepath.Join(cacheDir, "manifests")
+
+	require.NoError(t, os.MkdirAll(blobDir, 0o755))
+	require.NoError(t, os.MkdirAll(manifestsDir, 0o755))
+
+	for _, data := range []fakeRegistryData{
```

Comment:
maybe we could actually use image cache code itself to generate the cache, though we don't have a `registry:port` registry in CI

Reply:
or we could use `registry-gcr-io.ci.svc:5000` one of these in CI

Reply:
oh actually I can just use `ghcr.io:443`

Reply:
yeah, that might work